core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageOutputStream.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+final class StreamMessageOutputStream<T> extends OutputStream {
+
+    private final StreamMessage<T> source;
+    private final StreamWriter<T> writer;
+    private final Function<? super HttpData, ? extends T> httpDataConverter;
+    private final ByteBuf buffer = Unpooled.buffer();
+    private boolean transferred;
+    private boolean closed;
+
+    StreamMessageOutputStream(StreamMessage<T> source,
+                              StreamWriter<T> writer,
+                              Function<? super HttpData, ? extends T> httpDataConverter) {
+        requireNonNull(source, "source");
+        requireNonNull(writer, "writer");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        this.source = source;
+        this.writer = writer;
+        this.httpDataConverter = httpDataConverter;
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        ensureOpen();
+        buffer.writeByte(b);
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        ensureOpen();
+        buffer.writeBytes(b, off, len);
```

Comment:
Could we add configurable max buffer size and drain automatically when `buffer.readableBytes()` returns a greater value than that?